cpp/cmake_modules/SetupCxxFlags.cmake

```diff
@@ -274,29 +288,60 @@ if(BUILD_WARNING_FLAGS)
 endif(BUILD_WARNING_FLAGS)
 
 # Only enable additional instruction sets if they are supported
-if(CXX_SUPPORTS_AVX512 AND ARROW_SIMD_LEVEL STREQUAL "AVX512")
-  # skylake-avx512 consist of AVX512F,AVX512BW,AVX512VL,AVX512CD,AVX512DQ
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -march=skylake-avx512")
-elseif(CXX_SUPPORTS_AVX2 AND ARROW_SIMD_LEVEL STREQUAL "AVX2")
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -mavx2")
-elseif(CXX_SUPPORTS_SSE4_2 AND ARROW_SIMD_LEVEL STREQUAL "SSE4_2")
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -msse4.2")
+if(ARROW_USE_SIMD)
+  if(ARROW_SIMD_LEVEL STREQUAL "AVX512")
+    if(NOT CXX_SUPPORTS_AVX512)
+      message(FATAL_ERROR "AVX512 required but compiler doesn't support it.")
+    endif()
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_AVX512_FLAG}")
+    add_definitions(-DARROW_HAVE_AVX512 -DARROW_HAVE_AVX2)
+    add_definitions(-DARROW_HAVE_SSE4_2 -DARROW_HAVE_SSE2)
+  elseif(ARROW_SIMD_LEVEL STREQUAL "AVX2")
+    if(NOT CXX_SUPPORTS_AVX2)
+      message(FATAL_ERROR "AVX2 required but compiler doesn't support it.")
+    endif()
+    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_AVX2_FLAG}")
+    add_definitions(-DARROW_HAVE_AVX2)
+    add_definitions(-DARROW_HAVE_SSE4_2 -DARROW_HAVE_SSE2)
+  elseif(ARROW_SIMD_LEVEL STREQUAL "SSE4_2")
+    if(CXX_SUPPORTS_SSE4_2)
+      set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_SSE4_2_FLAG}")
+      add_definitions(-DARROW_HAVE_SSE4_2 -DARROW_HAVE_SSE2)
+    elseif(CXX_SUPPORTS_SSE2)
+      message(WARNING "Compiler doesn't support SSE4.2. Demote to SSE2.")
+      set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_SSE2_FLAG}")
+      add_definitions(-DARROW_HAVE_SSE2)
+    else()
+      message(STATUS "Compiler doesn't support SSE.")
+    endif()
+  endif()
 endif()
 
 if(CXX_SUPPORTS_ALTIVEC AND ARROW_ALTIVEC)
-  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} -maltivec")
+  set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_ALTIVEC_FLAG}")
 endif()
 
-if(CXX_SUPPORTS_ARMCRC)
-  if(CXX_SUPPORTS_ARMV8_CRC_CRYPTO)
-    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_ARMV8_CRC_CRYPTO_FLAG}")
-  else()
-    set(CXX_COMMON_FLAGS "${CXX_COMMON_FLAGS} ${ARROW_ARMV8_CRC_FLAG}")
+if(CXX_SUPPORTS_ARMV8)
+  if(NOT CXX_SUPPORTS_ARMV8_ARCH)
```

Comment:
I don't understand the difference between `CXX_SUPPORTS_ARMV8` and `CXX_SUPPORTS_ARMV8_ARCH`. Can you explain?

Reply:
Will be clear after refined with CMAKE_SYSTEM_PROCESSOR.

CXX_SUPPORTS_ARMV8 detects if compiler supports arm64 (-march=armv8-a)
CXX_SUPPORTS_ARMV8_ARCH detects if compiler supports arch requirement from cmake command line (-DARROW_ARMV8_ARCH=armv8.1-a+crypto)